### PR TITLE
Guest endpoint change

### PR DESCRIPTION
**Pull request by tomrf1**
Currently when we create a guest account, identity returns a guest token which we can use to create a password.
On the contributions thank you page we ask users to create a password.

This approach is now deprecated. We should stop using the guest token and instead make requests to `/guest?accountVerificationEmail=true`, which will automatically send an email asking the user to create a password.

#### New user:
Screen Shot 2021-09-01 at 14 06 57

#### Existing user, not signed in:
Screen Shot 2021-09-01 at 14 06 26

### Changes

1. Remove password creation component from contributions thank you page, and replace it with a message asking users to check their email.
2. Remove guest token logic from client.
3. Change support-frontend's backend to request `accountVerificationEmail=true`, and stop handling guest token.
4. Change payment-api to request `accountVerificationEmail=true`, and stop handling guest token.
5. On the client, use the userType (`new`,`guest`,`current`) to decide what to show on the thank you page. This involves putting it in session storage so that it's available after the paypal redirect.

### Deployment
payment-api must be deployed before support-frontend.
The guest token field is optional, so the client will not break when payment-api stops returning the token.

### CODE REVIEW

File: support-frontend/assets/helpers/forms/paymentIntegrations/oneOffContributions.js [hunk 1]
```diff
@@ -124,23 +124,12 @@ function paymentApiEndpointWithMode(url: string) {
 // Object is expected to have structure:
 // { type: "error", error: { failureReason: string } }, or
 // { type: "success", data: { currency: string, amount: number } }
-function paymentResultFromObject(
-  json: Object,
-  setGuestAccountCreationToken: (string) => void,
-  setThankYouPageStage: (ThankYouPageStage) => void,
```
Comment (tomrf1) [inline support-frontend/assets/helpers/forms/paymentIntegrations/oneOffContributions.js:130]: `ThankYouPageStage` is not actually used so I've removed this at the same time

File: support-frontend/assets/pages/contributions-landing/components/ContributionThankYou/ContributionThankYouSignUp.jsx [hunk 2]
```diff
@@ -1,228 +1,50 @@
...
-  password,
-  guestAccountCreationToken,
-  updatePassword,
-  setPasswordError,
-  passwordError,
-}: ContributionThankYouSignUpProps) => {
-  const [isExpanded, setIsExpanded] = useState(false);
-  const [hasBeenCompleted, setHasBeenInteractedWith] = useState(false);
-
+const ContributionThankYouSignUp = () => {
```
Comment (tomrf1) [inline support-frontend/assets/pages/contributions-landing/components/ContributionThankYou/ContributionThankYouSignUp.jsx:20]: This component used to ask the user to enter a password. Now it just tells them to check their email

File: support-payment-api/src/test/scala/backend/PaypalBackendSpec.scala [hunk 3]
```diff
@@ -19,7 +19,7 @@ import org.scalatestplus.mockito.MockitoSugar
 import services._
 import util.FutureEitherValues
 
-import scala.collection.JavaConverters._
```
Comment (tomrf1) [inline support-payment-api/src/test/scala/backend/PaypalBackendSpec.scala:22]: causes deprecated warnings

File: support-frontend/assets/pages/contributions-landing/contributionsLandingInit.js [hunk 4]
```diff
@@ -185,14 +186,6 @@ const init = (store: Store<State, Action, Function>) => {
 
   initialisePaymentMethods(state, dispatch);
 
-  // This will be in window.guardian if it has come from a PayPal one-off contribution,
-  // where it is returned by the Payment API to the backend, flashed into the session to preserve
-  // it through a serverside redirect, and then written into window.guardian on the thank-you page.
-  if (window.guardian.guestAccountCreationToken) {
-    dispatch(setGuestAccountCreationToken(window.guardian.guestAccountCreationToken));
-    dispatch(setThankYouPageStage('thankYouSetPassword'));
```
Comment (johnduffell) [inline support-frontend/assets/pages/contributions-landing/contributionsLandingInit.js:193]: does the thank you page just have one "default" stage now?

File: support-payment-api/src/main/scala/backend/PaypalBackend.scala [hunk 5]
```diff
@@ -69,13 +69,13 @@ class PaypalBackend(
...
 
           maybeEmail.foreach { email =>
-            getOrCreateIdentityIdFromEmail(email).foreach { identityIdWithGuestAccountCreationToken =>
-              postPaymentTasks(payment, email, identityIdWithGuestAccountCreationToken.map(_.identityId), capturePaymentData.acquisitionData, clientBrowserInfo)
+            getOrCreateIdentityIdFromEmail(email).foreach { identityId =>
+              postPaymentTasks(payment, email, identityId, capturePaymentData.acquisitionData, clientBrowserInfo)
             }
           }
 
           // The app doesn't need the guest account token in the response, because it has no 'set password' step after payment
```
Comment (johnduffell) [inline support-payment-api/src/main/scala/backend/PaypalBackend.scala:77]: is this comment still relevant? as it applies to everyone now

File: support-payment-api/src/main/scala/backend/PaypalBackend.scala [hunk 6]
```diff
@@ -142,7 +142,7 @@ class PaypalBackend(
   }
 
 
-  private def getOrCreateIdentityIdFromEmail(email: String): Future[Option[IdentityIdWithGuestAccountCreationToken]] =
+  private def getOrCreateIdentityIdFromEmail(email: String): Future[Option[Long]] =
```
Comment (johnduffell) [inline support-payment-api/src/main/scala/backend/PaypalBackend.scala:145]: identity id is surely an (opaque) String not a Long?

File: support-frontend/assets/pages/contributions-landing/contributionsLandingInit.js (hunk 4 above)
Comment (tomrf1) [reply support-frontend/assets/pages/contributions-landing/contributionsLandingInit.js:193]: yes, the ThankYouPageStage is a legacy thing that could have been removed previously. We used to have 2 pages, but now it's all on a single page and doesn't require this state transition.

File: support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.js [hunk 7]
```diff
@@ -339,31 +330,8 @@ function postRegularPaymentRequest(
...
 }
 
-function setPasswordGuest(
-  password: string,
-  guestAccountRegistrationToken: string,
-  csrf: CsrfState,
-): Promise<boolean> {
-
-  const data = { password, guestAccountRegistrationToken };
-  return logPromise(fetch(`${routes.contributionsSetPasswordGuest}`, requestOptions(data, 'same-origin', 'PUT', csrf)))
```
Comment (johnduffell) [inline support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.js:349]: if we no longer use the endpoint, will we get rid of it completely?

**Review by johnduffell — APPROVED** [review]
nice work, that's a lot of simplification which is great, I have left a couple of comments/questions for your perusal but nothing to stop it shipping 👍

File: support-frontend/assets/helpers/subscriptionsForms/submit.js [hunk 8]
```diff
@@ -217,8 +217,6 @@ function onPaymentAuthorised(
     data,
     abParticipations,
     csrf,
-    () => {},
```
Comment (rupertbates) [inline support-frontend/assets/helpers/subscriptionsForms/submit.js:220]: Nice to get rid of these horrors

File: support-frontend/assets/helpers/forms/paymentIntegrations/oneOffContributions.js [hunk 9]
```diff
@@ -151,11 +140,8 @@ const postToPaymentApi = (data: Object, path: string): Promise<Object> => fetchJ
 
 // Sends a one-off payment request to the payment API and standardises the result
 // https://github.com/guardian/payment-api/blob/master/src/main/resources/routes#L17
-const handleOneOffExecution = (result: Promise<Object>) => (
-  setGuestAccountCreationToken: (string) => void,
-  setThankYouPageStage: (ThankYouPageStage) => void,
-): Promise<PaymentResult> => logPromise(result)
-  .then(r => paymentResultFromObject(r, setGuestAccountCreationToken, setThankYouPageStage));
+const handleOneOffExecution = (result: Promise<Object>): Promise<PaymentResult> => logPromise(result)
```
Comment (rupertbates) [inline support-frontend/assets/helpers/forms/paymentIntegrations/oneOffContributions.js:143]: What is this change doing?

File: support-payment-api/src/main/scala/backend/SubscribeWithGoogleBackend.scala [hunk 10]
```diff
@@ -64,8 +64,8 @@ case class SubscribeWithGoogleBackend(databaseService: ContributionsStoreService
       _ => {
         logger.warn(s"unable to get identity id for email ${googleRecordPayment.email} with paymentId : ${googleRecordPayment.paymentId}, tracking acquisition anyway")
         handleContributionWithoutIdentity(googleRecordPayment, clientBrowserInfo).value
-      }, idWithGuestToken =>
```
Comment (rupertbates) [inline support-payment-api/src/main/scala/backend/SubscribeWithGoogleBackend.scala:67]: This backend could be deleted now couldn't it? We don't support Subscribe with Google any more

**Review by rupertbates — APPROVED** [review]
This turned into a bit of a monster didn't it?! Dom has asked if we can hold off from merging until Mahesh is back on Monday

File: support-payment-api/src/main/scala/backend/PaypalBackend.scala (hunk 6 above)
Comment (tomrf1) [reply support-payment-api/src/main/scala/backend/PaypalBackend.scala:145]: you're right, though payment-api has always decoded to Long! I'll leave this for a separate PR

File: support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.js (hunk 7 above)
Comment (tomrf1) [reply support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.js:349]: good spot, will do

File: support-payment-api/src/main/scala/backend/SubscribeWithGoogleBackend.scala (hunk 10 above)
Comment (tomrf1) [reply support-payment-api/src/main/scala/backend/SubscribeWithGoogleBackend.scala:67]: yes! I've been meaning to do this for about a year now

**Review by johnduffell — APPROVED** [review]
thanks for tidying up the unused endpoint also 👍

**Pull request merged**